### PR TITLE
avoid authN/authZ bypass for inbound protocol sniffing

### PR DESCRIPTION
**Pull request by yxue**
avoid authz bypass when protocol sniffing is enabled on inbound

Fix https://github.com/istio/istio/issues/16214

### CODE REVIEW

**Comment by yxue** [comment]
@rshriram the PR is intended to fix the authn bypass. I haven't added test. Before that, could you take a look at the PR? Should I move the fix to Pilot or keep that in plugin?

**Comment by rshriram** [comment]
Per offline discussion, lets keep this in the plugin. Else we would end up having a lot of dependencies on the plugins. The plugins were designed to mutate the filter chains. So what is being done here is good [i.e. if authN filter exists, use http]. In fact, if you want to take this idea further, you can even scan the list of virtual services for the sniffed port. If you have virtual services using http only, then you can simply do http listener.

**Comment by rshriram** [comment]
and yes, please add a test to make sure this is working properly.. in the e2e. may be make one of the ports as auto?

File: pilot/pkg/networking/plugin/authn/authentication.go [hunk 1]
```diff
@@ -67,18 +68,32 @@ func buildFilter(in *plugin.InputParams, mutable *plugin.MutableObjects) error {
...
 				mutable.FilterChains[i].HTTP = append(mutable.FilterChains[i].HTTP, filter)
+				hasAuthnFilter = true
 			}
+			HTTPFilterChains = append(HTTPFilterChains, mutable.FilterChains[i])
+			listenerHTTPFilterChain = append(listenerHTTPFilterChain, mutable.Listener.FilterChains[i])
 		}
 	}
 
+	if hasAuthnFilter && in.ListenerProtocol == plugin.ListenerProtocolAuto {
+		mutable.FilterChains = HTTPFilterChains
```
Comment (yangminzhu) [inline pilot/pkg/networking/plugin/authn/authentication.go]: Could you explain a little bit and also add a comment how is this working?

**Comment by yxue** [comment]
Find some issues when testing. If there is a mesh wise authN, all the listeners on unnamed ports will only have HTTP filter chain.

**Comment by rshriram** [comment]
add the code to authZ and not authN. And i think this code needs to be flag guarded if people are okay with breaking backward compatibility [when there are global authZ rules].

**Pull request closed without merging**